### PR TITLE
Bury bip9 deployments

### PR DESCRIPTION
**Pull request by jnewbery**
This hardcodes CSV and segwit activation heights, similar to the BIP 90 buried deployments for BIPs 34, 65 and 66.

Both CSV and segwit have been active for over 8 months. Hardcoding the activation height is a code simplification, makes it easier to understand segwit activation status, and reduces technical debt.

### CODE REVIEW

**Comment by jtimon** [comment]
This will need rebase after https://github.com/bitcoin/bitcoin/pull/11739 , won't it?

**Comment by jnewbery** [comment]
> This will need rebase after #11739 , won't it?

There are no dependencies between the two, but you're correct that there are some minor conflicts in the implementations. I'm happy to rebase if 11739 is merged first.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -1074,54 +1074,47 @@ UniValue verifychain(const JSONRPCRequest& request)
...
 
-static UniValue SoftForkDesc(const std::string &name, int version, CBlockIndex* pindex, const Consensus::Params& consensusParams)
-{
     UniValue rv(UniValue::VOBJ);
-    rv.push_back(Pair("id", name));
-    rv.push_back(Pair("version", version));
-    rv.push_back(Pair("reject", SoftForkMajorityDesc(version, pindex, consensusParams)));
-    return rv;
+    rv.pushKV("type", "buried");
+    rv.pushKV("type", "buried");
```
Comment (sipa) [inline src/rpc/blockchain.cpp]: Duplicated line?

Comment (promag) [reply src/rpc/blockchain.cpp]: Yes? :trollface:

File: src/chainparams.cpp [hunk 2]
```diff
@@ -59,6 +59,11 @@ void CChainParams::UpdateVersionBitsParameters(Consensus::DeploymentPos d, int64
     consensus.vDeployments[d].nTimeout = nTimeout;
 }
 
+void CChainParams::UpdateSegwitHeight(const int& height)
```
Comment (promag) [inline src/chainparams.cpp]: Any good reason for reference?

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (jnewbery) [reply src/rpc/blockchain.cpp]: Fixed

File: src/chainparams.cpp (hunk 2 above)
Comment (jnewbery) [reply src/chainparams.cpp]: None. Removed.

File: src/consensus/params.h [hunk 3]
```diff
@@ -58,6 +56,10 @@ struct Params {
     int BIP65Height;
     /** Block height at which BIP66 becomes active */
     int BIP66Height;
+    /** Block height at which CSV becomes active */
+    int CSVHeight;
+    /** Block height at which Segwit becomes active */
```
Comment (Empact) [inline src/consensus/params.h]: Is it helpful to continue referencing the BIP numbers in these comments? I would guess so.

Comment (jnewbery) [reply src/consensus/params.h]: Sure. Comments added.

Comment (dcousens) [reply src/consensus/params.h]: Why not refer to the BIP numbers solely instead of alternating?

Comment (MarcoFalke) [reply src/consensus/params.h]: the segwit deployment covered more than one bip

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -1160,36 +1155,28 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
...
-            "  \"softforks\": [                (array) status of softforks in progress\n"
-            "     {\n"
-            "        \"id\": \"xxxx\",           (string) name of softfork\n"
-            "        \"version\": xx,          (numeric) block version\n"
-            "        \"reject\": {             (object) progress toward rejecting pre-softfork blocks\n"
-            "           \"status\": xx,        (boolean) true if threshold reached\n"
+            "  \"chainwork\": \"xxxx\"     (string) total amount of work in active chain, in hexadecimal\n"
+            "  \"pruned\": xx,             (boolean) if the blocks are subject to pruning\n"
+            "  \"pruneheight\": xxxxxx,    (numeric) lowest-height complete block stored\n"
+            "  \"softforks\": {               (object) status of softforks\n"
```
Comment (jamesob) [inline src/rpc/blockchain.cpp]: We don't seem to have any testing for the contents of this key - is this the case? Happy to either write some for this PR or file a follow-up.

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -1160,36 +1155,28 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "  \"difficulty\": xxxxxx,         (numeric) the current difficulty\n"
             "  \"mediantime\": xxxxxx,         (numeric) median time for the current best block\n"
             "  \"verificationprogress\": xxxx, (numeric) estimate of verification progress [0..1]\n"
-            "  \"initialblockdownload\": xxxx, (bool) (debug information) estimate of whether this node is in Initial Block Download mode.\n"
```
Comment (jamesob) [inline src/rpc/blockchain.cpp]: Curious why removal of this key and others (e.g. `size_on_disk`) is bundled with this changeset. Seems like RPC interface changes should be handled separately.

**Review by jamesob — COMMENTED** [review]
Concept ACK https://github.com/bitcoin/bitcoin/pull/12360/commits/32cbe4828d7d22f9c88ce9ea27c69cd16cc8ebbe

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -1074,54 +1074,46 @@ UniValue verifychain(const JSONRPCRequest& request)
...
-{
     UniValue rv(UniValue::VOBJ);
-    rv.push_back(Pair("id", name));
-    rv.push_back(Pair("version", version));
-    rv.push_back(Pair("reject", SoftForkMajorityDesc(version, pindex, consensusParams)));
-    return rv;
+    rv.pushKV("type", "buried");
+    // getblockchaininfo reports the softfork as active from when the chain height is
+    // one below the activation height
+    rv.push_back(Pair("active", chainActive.Tip()->nHeight + 1 >= height));
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/rpc/blockchain.cpp]: Requires #12193

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (jnewbery) [reply src/rpc/blockchain.cpp]: You're right. `rpc_blockchain.py` should be updated. If you're happy to contribute a commit to add tests, I can include it in this PR.

File: src/rpc/blockchain.cpp (hunk 5 above)
Comment (jnewbery) [reply src/rpc/blockchain.cpp]: Oops. Bad rebase. Should be fixed now.

File: src/rpc/blockchain.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/rpc/blockchain.cpp]: ah. It's just been merged. Will rebase.

**Comment by jnewbery** [comment]
Rebased now that #12193 has been merged.

ACK @jamesob's 0be1f9d8a6bb2b8c47d6816b46d6764aee3cc06e commit.

**Comment by jamesob** [comment]
Biased ACK https://github.com/bitcoin/bitcoin/pull/12360/commits/0be1f9d8a6bb2b8c47d6816b46d6764aee3cc06e

**Comment by jtimon** [comment]
utACK 0be1f9d8a6bb2b8c47d6816b46d6764aee3cc06e

**Comment by jnewbery** [comment]
rebased

**Comment by MarcoFalke** [comment]
Needs rebase to fix travis (sorry)

**Comment by jamesob** [comment]
Rebased.

**Comment by jnewbery** [comment]
Rebased

File: src/init.cpp [hunk 7]
```diff
@@ -1158,6 +1159,21 @@ bool AppInitParameterInteraction()
             }
         }
     }
+
```
Comment (jtimon) [inline src/init.cpp]: This be done directly in CRegtestParams's constructor, that way, the function UpdateSegwitHeight would not be needed.

**Review by jtimon — CHANGES_REQUESTED** [review]

File: src/init.cpp (hunk 7 above)
Comment (jnewbery) [reply src/init.cpp]: Seems like a good idea, but segwit height is also updated in one of the unit tests:

https://github.com/bitcoin/bitcoin/pull/12360/files#diff-d5ba361c5f8be78eb4cc0c787c1fc78eR118

I think we need the `UpdateSegwitHeight()` function call there.

Comment (jtimon) [reply src/init.cpp]: I guess we still need UpdateSegwitHeight() then, or perhaps we can pass the other args (instead of g_args, or rather g_args plus "-segwitheight=432") to SelectParams in https://github.com/bitcoin/bitcoin/pull/12360/files#diff-d5ba361c5f8be78eb4cc0c787c1fc78eR54 ?
I guess not all BasicTestingSetup want 432 there...

Comment (jnewbery) [reply src/init.cpp]: I think I'll leave it as is for now. It may be possible to tidy this up in a future commit.

**Comment by jnewbery** [comment]
Rebased

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
#11739 is merged. Rebased

**Comment by MarcoFalke** [comment]
Tend to NACK. Just because buried deployments can be done safely doesn't mean they should be done for the sole purpose of code refactoring without substantial motivation, advantages or simplifications. Since this is refactoring consensus code and also changing the consensus rules, I think a writeup on the motivation and tradeoffs is required. Note that none of the statements of the motivation given in BIP90 are applicable here, so I doubt that the overall gain (if any) is worth the time spent on this (code/discussions/formal writeups/review/...). I'd rather keep the deployments as versionbits deployments and instead treat `IsWitnessEnabled` as `true` for modules that don't need to know when the witness commitment started to be a requirement for blocks containing witness transactions. I'd guess compact block relay qualifies for this, since it only relays recent blocks. Alternatively the "mining" code, which only makes sense to call on chains where `IsWitnessEnabled` is already `true`.

**Comment by MarcoFalke** [comment]
Potentially also the mempool, after which #12124 could be rebased and reopened.

**Comment by jnewbery** [comment]
@MarcoFalke - I disagree. Once a softfork is deployed and is sufficiently buried, the activation mechanism is purely academic. Removing complexity and improving the simplicity of the code is motivation enough.

> I think a writeup on the motivation and tradeoffs is required

Yes - I intend to send a writeup to the mailing list if this PR is merged.

As a meta-point, although review and feedback is always welcome, concept NACKs are much more useful early after a PR has been opened. This PR is a rebase of #11398, which was opened 7 months ago and has now been rebased ~10 times (including once requested by [you](https://github.com/bitcoin/bitcoin/pull/11398#issuecomment-343550044)). This PR and its antecedent have already been ACKed or concept ACKed by 5 contributors. If your objection is time spent on discussion/review, then that ship has already sailed :)

**Comment by MarcoFalke** [comment]
> Removing complexity and improving the simplicity of the code is motivation enough.

I am not aware of any plans to get rid of the concept of versionbits-deployments, so the "simplifications and improvements" of the code are limited to that module of the code and don't have beneficial side-effects or simplifications (such as BIP-90).

I am not saying that the changes can not be made, but rather that there is no need to to them right now. It is good to keep this patch in mind and there is no harm in having it in a rebased form... My Concept NACK is more a -0.

Please add some minimal motivation to the opening comment, if you plan to keep on curating this pull. I'd assume for new developers those changes to the consensus code are hard to follow in the absence of any motivation and the presence of merely a blob of code.

File: src/init.cpp [hunk 8]
```diff
@@ -1149,6 +1150,21 @@ bool AppInitParameterInteraction()
...
+    if (gArgs.IsArgSet("-segwitheight")) {
+        if (!chainparams.MineBlocksOnDemand()) {
+            return InitError("Segwit activation parameters may only be overridden on regtest.");
+        }
+        int64_t height = gArgs.GetArg("-segwitheight", chainparams.GetConsensus().SegwitHeight);
+        if (height < -1 || height >= std::numeric_limits<int>::max()) {
+            return InitError(strprintf("Activation height %ld for segwit is out of valid range. Use -1 to disable segwit.", height));
+        }
+        else if (height == -1) {
+            LogPrintf("Segwit disabled for testing\n");
```
Comment (MarcoFalke) [inline src/init.cpp]: It seems odd to add code just for the tests, that has no application outside of regtest. I'd prefer if the code paths in the tests that test segwit can be disabled are just removed. Afterward this code block and all other dead code (`IsScriptWitnessEnabled`, ...) can be discarded.

**Comment by sdaftuar** [comment]
In my view, segwit activation at _some_ height is the consensus that the software assumes -- given that our own wallet now generates segwit transactions by default it would be internally inconsistent to tolerate being on a chain on which segwit might never be active.  Moreover, we're contemplating merging a change in #13120 that would change the mempool logic to always accept segwit transactions, even if we were on a chain on which segwit was not active.

I do agree with @MarcoFalke's comment that this is not a necessary change; however I think it is still a desirable one -- a height check seems simpler than VersionBits-based logic, even if making that change touches a bunch of code.  For instance, future code readers interested in segwit logic should not have to wonder under what circumstances the VersionBits calls might not return `THRESHOLD_ACTIVE`.  Further, I think replacing the VB logic with a height check also more accurate captures the uncertainty around the activation, where it was unclear whether segwit activated according to the rules specified in BIP 141, BIP 91, or BIP 148 (or something else?).  Maintaining the software's BIP 141 activation logic is an unnecessary and somewhat arbitrary burden, even if so far it has been a light one.

> I am not aware of any plans to get rid of the concept of versionbits-deployments, so the "simplifications and improvements" of the code are limited to that module of the code and don't have beneficial side-effects or simplifications (such as BIP-90).

I disagree with the second half that sentence -- I think that the circumstances around whether segwit is or might be active do concern other parts of the code (such as the wallet, mempool, and mining, which must make choices around what transactions to generate, accept or select) and therefore it's easier to reason about the possibilities if we remove chain-specific calculations and replace them with the height checks proposed here.

In summary, I think replacing the activation logic with a height check makes the code more internally consistent and simpler to understand and reason about, so I think this is worth doing.  Concept ACK.

File: src/init.cpp (hunk 8 above)
Comment (jnewbery) [reply src/init.cpp]: I completely agree. The only reason the 'Segwit disabled' code here and elsewhere exists is for `p2p_segwit.py`, where it's used to test whether GBT doesn't return a default witness commitment if segwit is disabled. That's completely circular - we're adding behaviour into the product so we can test whether that behaviour exists.

I'd like to completely overhaul p2p_segwit.py and remove this special casing, but in a future PR. This PR is simply maintaining the existing functionality of setting `-vbparams=segwit:0:0`.

File: test/functional/feature_segwit.py [hunk 9]
```diff
@@ -42,9 +42,9 @@ def set_test_params(self):
         self.setup_clean_chain = True
         self.num_nodes = 3
         # This test tests SegWit both pre and post-activation, so use the normal BIP9 activation.
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply test/functional/feature_segwit.py]: fixed (here and in `p2p_segwit.py`).

**Comment by jnewbery** [comment]
Fixed @MarcoFalke's comment and tidied up p2p_segwit.py a little.

**Comment by MarcoFalke** [comment]
Could the rpc changes potentially be split up into a separate pull request, to not distract from the consensus changes?

**Comment by jnewbery** [comment]
> Could the rpc changes potentially be split up into a separate pull request, to not distract from the consensus changes?

It's possible, but this PR already has two ACKs and a Concept ACK. I think that splitting it up would be generating more work for reviewers.

**Comment by jnewbery** [comment]
rebased

**Comment by MarcoFalke** [comment]
> It's possible, but this PR already has two ACKs and a Concept ACK. I think that splitting it up would be generating more work for reviewers.

The ACK are outdated and would need to be re-ACKed on the new commits anyway.

**Comment by jnewbery** [comment]
Fixed failling p2p_segwit.py

**Comment by jnewbery** [comment]
This is a revamped version of @jl2012's #11398, which had concept ACKs from @gmaxwell @dcousens and @sdaftuar . This PR also has ACKs from @jamesob and @jtimon .

Are people still interested in this? I haven't had any feedback for the last three rebases (except @MarcoFalke's NACKs), and I'd rather not continue rebasing if people have cooled on this idea.

**Comment by jnewbery** [comment]
Closing - 6 concept ACKs but no review so I guess this isn't worth rebasing.
